### DISCUSSION TITLE
Streaming LOBs, writing over 1gb of data to table CLOB column with JSON constraint?

### DISCUSSION THREAD

**Discussion by Christopher-Gentle** [oracle/python-oracledb#113] [Q&A]
Hello,

Using Oracle 19c,
Is it possible to stream data from a >1GB JSON file to a table with a CLOB column that has a JSON check constraint?

I looked at the example below, it seems they do not have the JSON check constraint. I tried to adapt it to my requirement, but it fails with a JSON validation error.

```python
id_val = 9
lob_var = cursor.var(oracledb.DB_TYPE_BLOB)
cursor.execute("""
        insert into lob_tbl (id, b)
        values (:1, empty_blob())
        returning b into :2""", [id_val, lob_var])
blob, = lobVar.getvalue()
offset = 1
num_bytes_in_chunk = 65536
with open("image.png", "rb") as f:
    while True:
        data = f.read(num_bytes_in_chunk)
        if data:
            blob.write(data, offset)
        if len(data) < num_bytes_in_chunk:
            break
        offset += len(data)
connection.commit()
```

**Comment by anthony-tuininga** [comment]
Interesting! Can you share the code you actually used that got the JSON validation error? Specifically I'm interested to know *when* you get the error.

**Reply by Christopher-Gentle** [reply]
Don't mind the OracleHook, just using Airflow with cx_Oracle.

clob.write(data, offset)
cx_Oracle.DatabaseError: ORA-40441: JSON syntax error

```python
def cx_lob_ex():
    conn = OracleHook(get_oracle_conn_id_by_env()).get_conn()
    with conn.cursor() as cursor:
        lob_var = cursor.var(cx_Oracle.DB_TYPE_CLOB)
        cursor.execute(
            """
            insert into CX_EAMPLE_TABLE (JSON_DOC)
            values (empty_clob())
            returning JSON_DOC into :1""",
            [lob_var],
        )
        (clob,) = lob_var.getvalue()
        offset = 1
        num_bytes_in_chunk = 65536
        with open("/opt/airflow/data/example.json", "rb") as f:
            while True:
                data = f.read(num_bytes_in_chunk)
                if data:
                    clob.write(data, offset)
                if len(data) < num_bytes_in_chunk:
                    break
                offset += len(data)
        conn.commit()
```

**Reply by anthony-tuininga** [reply]
Thanks. Can you also supply the traceback that you are getting when you get the error?

**Reply by Christopher-Gentle** [reply]
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/decorators/base.py", line 179, in execute
    return_value = super().execute(context)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/operators/python.py", line 171, in execute
    return_value = self.execute_callable()
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/operators/python.py", line 189, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
  File "/opt/airflow/dags/etl_tg.py", line 86, in cx_oracle_ex
    clob.write(data, offset)
cx_Oracle.DatabaseError: ORA-40441: JSON syntax error

**Reply by anthony-tuininga** [reply]
Can you try making the constraint deferrable? That should defer the checking of the JSON until the commit is performed; otherwise, on every write the data must be valid JSON -- which of course isn't possible!

**Reply by Christopher-Gentle** [reply]
The JSON file inserts fine with DBMS_LOB.loadclobfromfile. I just don't have the ability to get the file to the host Oracle server so it's not practical in my use-case.

I'm not entirely sure what the write method above is doing to debug this. I suspect it's trying to incrementally insert the JSON data so it's malformed and that is what is causing the error?

The JSON file is formatted properly but it's over 1GB. The table is designed to an entire JSON file into one row/one column (JSON_DOC). I was maybe wondering if there was a way to stream the data to a temporary CLOB on the host Oracle DB, and insert it using a PL/SQL block? Thus, avoiding memory issues on the client side?

Also, I believe there is a slight typo in the example:
`blob, = lobVar.getvalue()`
Should be:
`blob, = lob_var.getvalue()`

but I could be wrong.

Thanks for any help!

**Reply by Christopher-Gentle** [reply]
> Can you try making the constraint deferrable? That should defer the checking of the JSON until the commit is performed; otherwise, on every write the data must be valid JSON -- which of course isn't possible!

Good idea, let me check if I can do that, I will get back to you.

Thanks!

**Reply by Christopher-Gentle** [reply]
Okay, re-created the constraint:
DEFERRABLE INITIALLY DEFERRED

In SQL Developer, this does work now:
insert into CX_EXAMPLE_TABLE (JSON_DOC) values ('{"data": [{"Id": 1, "Loan": "5XXXX", "NextPrincipalBillDate": "2017-06-01T00:00:00",')
It will say 1 row created until I commit at which point it rolls back as expected because malformed JSON.

I am able to partially add JSON in SQL Developer but it is still failing with cx_Oracle with the same error and traceback.
clob.write(data, offset)
cx_Oracle.DatabaseError: ORA-40441: JSON syntax error

**Reply by Christopher-Gentle** [reply]
Yeah it's an issue related to the IS JSON constraint. I removed it completely from the table and was able to run the sample code.
I don't think this is the expected behavior?

**Reply by anthony-tuininga** [reply]
I'm not sure. I can ask internally and let you know if I get any useful information.

**Reply by anthony-tuininga** [reply]
One thought I had which you can try:

```python
blob.open()
try:
    offset = 1
    num_bytes_in_chunk = 65536
    with open("image.png", "rb") as f:
        while True:
            data = f.read(num_bytes_in_chunk)
            if data:
                blob.write(data, offset)
            if len(data) < num_bytes_in_chunk:
                break
            offset += len(data)
finally:
   blob.close()
```

This may solve the problem for you? If it does, please advise! We can add some notes in the documentation to let future readers avoid the pain you have been experiencing!

**Reply by Christopher-Gentle** [reply]
> One thought I had which you can try:
> 
> ```python
> blob.open()
> try:
>     offset = 1
>     num_bytes_in_chunk = 65536
>     with open("image.png", "rb") as f:
>         while True:
>             data = f.read(num_bytes_in_chunk)
>             if data:
>                 blob.write(data, offset)
>             if len(data) < num_bytes_in_chunk:
>                 break
>             offset += len(data)
> finally:
>    blob.close()
> ```
> 
> This may solve the problem for you? If it does, please advise! We can add some notes in the documentation to let future readers avoid the pain you have been experiencing!

Unfortunately, the same result.

clob.write(data, offset)
cx_Oracle.DatabaseError: ORA-40441: JSON syntax error

I'm not really sure where it is coming from, I suspect a commit is somewhere in the "write" method.

**Reply by anthony-tuininga** [reply]
Interesting. I'll have to see if I can try this myself. Does it fail in the same way even if the data is less than 1 GB (but is still written in chunks)?

**Reply by Christopher-Gentle** [reply]
yes. I have tested it with both large and small datasets.
The only way it will work like this is if I disable/remove the constraint.

If I was going to go that route -- I would just invoke sqlldr (constraints have to be disabled to load JSON anyway).
I was hoping to centralize all of my Oracle operations in cx_Oracle.

**Reply by anthony-tuininga** [reply]
Thanks for the confirmation. This may be a bug. Can you indicate which server version you are using? Also, do you get the same results in both thin and thick modes?

**Reply by Christopher-Gentle** [reply]
Oracle Database 19c Enterprise Edition Release 19.0.0.0.0
cx_Oracle version 8.3.0

I have not tried to switch between modes, let me look into that.
Thanks!

**Reply by Christopher-Gentle** [reply]
I tried both modes, no luck.
It doesn't even like the constraint being disabled. It has to be completely removed.

**Reply by Christopher-Gentle** [reply]
I did a bit more digging, seems to be an Oracle bug and not a cx_Oracle bug. Once you add a IS_JSON constraint, even if it's disabled, it will throw a JSON syntax error when trying to use DBMS_LOB.write IF any chunk of data is not completely intact JSON. The only work-around I have found is 1) removing the constraint, 2) disable constraint, use sqlldr to load the data, finally, enable the constraint. I'm curious what sqlldr is using if not DBMS_LOB.
